R/lazyframe-frame.R

```diff
@@ -358,12 +759,2052 @@ lazyframe__slice <- function(offset, length = NULL) {
+    by <- parse_into_list_of_expressions(...)
+    self$`_ldf`$join_where(other$`_ldf`, by, suffix)
+  })
+}
+
+#' Unpivot a LazyFrame from wide to long format
+#'
+#' This function is useful to massage a LazyFrame into a format where one or
+#' more columns are identifier variables (`index`) while all other columns,
+#' considered measured variables (`on`), are “unpivoted” to the row axis
+#' leaving just two non-identifier columns, "variable" and "value".
+#'
+#' @inheritParams rlang::check_dots_empty0
+#' @param on Values to use as identifier variables. If `value_vars` is
+#' empty all columns that are not in `id_vars` will be used.
+#' @param index Columns to use as identifier variables.
+#' @param variable_name Name to give to the new column containing the names of
+#' the melted columns. Defaults to "variable".
+#' @param value_name Name to give to the new column containing the values of
+#' the melted columns. Defaults to `"value"`.
+#'
+#' @inherit as_polars_lf return
+#'
+#' @examples
+#' lf <- pl$LazyFrame(
+#'   a = c("x", "y", "z"),
+#'   b = c(1, 3, 5),
+#'   c = c(2, 4, 6)
+#' )
+#' lf$unpivot(index = "a", on = c("b", "c"))$collect()
+lazyframe__unpivot <- function(
+    on = NULL,
+    ...,
+    index = NULL,
+    variable_name = NULL,
+    value_name = NULL) {
+  wrap({
+    check_dots_empty0(...)
+    if (!is.null(on)) {
+      on <- parse_into_list_of_expressions(!!!on)
+    }
+    if (!is.null(index)) {
+      index <- parse_into_list_of_expressions(!!!index)
+    }
+    self$`_ldf`$unpivot(on, index, value_name, variable_name)
+  })
+}
+
+#' Rename column names
+#'
+#' @param mapping Either a function that takes a character vector as input and
+#' returns one as input, or a named list where names are old column names and
+#' values are the new ones.
+#' @param ... <[`dynamic-dots`][rlang::dyn-dots]> If `mapping` is missing,
+#' those values are used.
+#' @param strict Validate that all column names exist in the current schema,
+#' and throw an error if any do not. (Note that this parameter is a no-op when
+#' passing a function to `mapping`).
+#'
+#' @details
+#' If existing names are swapped (e.g. 'A' points to 'B' and 'B' points to
+#' 'A'), polars will block projection and predicate pushdowns at this node.
+#'
+#' @inherit as_polars_lf return
+#' @examples
+#' lf <- pl$LazyFrame(
+#'   foo = 1:3,
+#'   bar = 6:8,
+#'   ham = letters[1:3]
+#' )
+#'
+#' lf$rename(foo = "apple")$collect()
+#'
+#' lf$rename(
+#'   \(column_name) paste0("c", substr(column_name, 2, 100))
+#' )$collect()
+lazyframe__rename <- function(mapping, ..., strict = TRUE) {
+  wrap({
+    if (!missing(mapping) && is_function(mapping)) {
+      check_dots_empty0(...)
+      self$select(pl$all()$name$map(mapping))
+    } else {
+      if (missing(mapping) || !is.list(mapping)) {
+        mapping <- list2(...)
+      }
+      existing <- names(mapping)
+      new <- unlist(mapping)
+      self$`_ldf`$rename(existing, new, strict)
+    }
+  })
+}
```

Comment:
At first glance, this function looks like something that should be rewritten.

Reply:
This is quite similar to the python implementation: https://github.com/pola-rs/polars/blob/87feed72585eff5acf3defb7f81029123d5cba68/py-polars/polars/lazyframe/frame.py#L5274-L5279

```
        if callable(mapping):
            return self.select(F.all().name.map(mapping))
        else:
            existing = list(mapping.keys())
            new = list(mapping.values())
            return self._from_pyldf(self._ldf.rename(existing, new, strict))
```
Since we want to use `mapping` if it's provided and `...` if `mapping` is absent we're going to have something a bit more complicated. What do you think about something like this?

```r
lazyframe__rename <- function(mapping, ..., .strict = TRUE) {
  wrap({
    if (!missing(mapping)) {
      check_dots_empty0(...)
      if (is_function(mapping)) {
        self$select(pl$all()$name$map(mapping))
      } else {
        abort("`mapping` must be a function when provided.")
      }
    } else {
      # Have a check_dots_named() function?
      mapping <- list2(...)
      existing <- names(mapping)
      new <- unlist(mapping)
      self$`_ldf`$rename(existing, new, .strict)
    }
  })
}
```

Reply:
In my opinion, functions that accept Python dict, such as this `rename` and `cast`, which is already in the main branch, should use dynamic dots instead of dict.

https://github.com/eitsupi/neo-r-polars/blob/365a66965243cfeb3909f81fa576b28aa01cd102/R/dataframe-frame.R#L427-L447

In any case, since the map_name function does not yet exist, I believe it is a TODO if a function is received.

Reply:
> In my opinion, functions that accept Python dict, such as this `rename` and `cast`, which is already in the main branch, should use dynamic dots instead of dict.

Agreed, the only tricky part here is that it can also accept a function. What do you think of the following?

```r
lazyframe__rename <- function(..., .strict = TRUE) {
  wrap({
    mapping <- list2(...)
    if (length(mapping) == 1 && is_function(mapping)) {
      check_dots_empty0(...)
      # TODO: this requires $name$map()
      return(self$select(pl$all()$name$map(mapping)))
    }
    existing <- names(mapping)
    new <- unlist(mapping)
    self$`_ldf`$rename(existing, new, .strict)
  })
}
```
I have added a TODO meanwhile.

Reply:
> What do you think of the following?

Thank you. I would prefer something like that. (Just wondering if a type check on the `mapping` would be necessary?)

Reply:
Done in [`25dcda8` (#31)](https://github.com/eitsupi/neo-r-polars/pull/31/commits/25dcda8db9289e19a970d4abf2722f35cb9d7c16). 

> Just wondering if a type check on the mapping would be necessary?

We could add something like 

```r
if (!all(vapply(mapping, is_character, logical(1)))) {
      abort("`...` only accepts a function or named characters.")
    }
```
although I'm not very happy with the wording.

Reply:
Since this is a frequent pattern, you can probably use `is_list_of_string()` to check.

Reply:
Done